src/IO/LimitSeekableReadBuffer.cpp

```diff
@@ -57,18 +66,16 @@ bool LimitSeekableReadBuffer::nextImpl()
         return false;
     }
 
-    /// Adjust the size of the buffer (we don't allow to read more than `max_offset - min_offset`).
-    off_t size = in->buffer().size();
-    size = std::min(size, max_offset - in->getPosition());
+    /// Adjust the beginning and the end of the working buffer (we don't allow to read more than `max_offset - min_offset`).
+    auto * begin = in->buffer().begin();
+    auto * end = in->buffer().end();
 
-    if (!size || (static_cast<off_t>(in->offset()) >= size))
-    {
-        /// Limit reached.
-        set(in->position(), 0);
-        return false;
-    }
+    begin = std::max(begin, in->position() - offset_after_min);
+    end = std::min(end, in->position() + available_before_max);
+
+    BufferBase::set(begin, end - begin, in->position() - begin);
+    chassert(position() == in->position() && available());
 
-    BufferBase::set(in->buffer().begin(), size, in->offset());
```

Comment:
Here the calculation was wrong. It started reading from the correct position but both the beginning and the end of the buffer were not adjusted correctly. As a result we correctly assigned the Content-Length for reading the appended file after the matched prefix, but passed a read buffer which couldn't read anything. As a result AWS SDK couldn't read anything from our read buffer and failed by timeout all the time.